File: Valid_Sudoku.cpp
/* Determine if a Sudoku is valid, according to: Sudoku Puzzles - The Rules.

The Sudoku board could be partially filled, where empty cells are filled with the character '.'.


A partially filled sudoku which is valid. */
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
         //check each row
         for(int i = 0; i < 9; i++){
             vector<int> numCount(9,0);
             for(int j = 0; j < 9; j++){
                 if(isdigit(board[i][j])){
                     numCount[board[i][j] - '1']++;
                     if(numCount[board[i][j] - '1'] > 1){
                         return false;
                     }
                 }
             }
         }
         // check each col
         for(int i = 0; i < 9; i++){
             vector<int> numCount(9,0);
             for(int j = 0; j < 9; j++){
                 if(isdigit(board[j][i])){
                     numCount[board[j][i] - '1']++;
                     if(numCount[board[j][i] - '1'] > 1){
                         return false;
                     }
                 }
             }
         }
         // check each block
         for(int i = 0; i < 3; i++){
             for(int j = 0; j < 3; j++){
                 vector<int> numCount(9,0);
                 for(int k = i*3; k < i*3 + 3; k++){
                     for(int kk = j*3; kk < j*3 + 3; kk++){
                         if(isdigit(board[k][kk])){
                             numCount[board[k][kk] - '1']++;
                             if(numCount[board[k][kk] - '1'] > 1){
                                 return false;
                             }
                         }
                     }
                 }
             }
         }
         return true;
    }
};

File: python/Valid_Sudoku.py
"""
Determine if a Sudoku is valid, according to: Sudoku Puzzles - The Rules.

The Sudoku board could be partially filled, where empty cells are filled with the character '.'.


A partially filled sudoku which is valid.
"""
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        # check each row
        for row in range(0,9):
            num = [0] * 9
            for col in range(0,9):
                if board[row][col].isdigit():
                    num[int(board[row][col]) - 1] += 1
                    if num[int(board[row][col]) - 1] > 1 : return False
        for col in range(0,9):
            num = [0] * 9
            for row in range(0,9):
                if board[row][col].isdigit():
                    num[int(board[row][col]) - 1] += 1
                    if num[int(board[row][col]) - 1] > 1 : return False


        for brow in range(0,3):
            for bcol in range(0,3):
                num = [0] * 9
                for k in range(0,3):
                    for kk in range(0,3):
                        if board[brow * 3 + k][bcol *3 + kk].isdigit():
                            num[int(board[brow * 3 + k][bcol *3 + kk]) - 1] += 1
                            if num[int(board[brow * 3 + k][bcol *3 + kk]) - 1] > 1 : return False
        return True       
